src/client/array/dc_array.c

```diff
@@ -2373,12 +2316,16 @@ adjust_array_size_cb(tse_task_t *task, void *data)
 			if (dkey_val == 0)
 				continue;
 			/*
-			 * Punch the entire dkey since it's in a higher dkey
-			 * group than the intended size.
+			 * The dkey is higher than the adjustded size so we could punch it here.
+			 * But it's better to punch the extent so that the max_write for the object
+			 * doesn't get lost by aggregation.
 			 */
-			D_DEBUG(DB_IO, "Punching key: "DF_U64"\n", dkey_val);
-			rc = punch_key(args->oh, args->th, dkey_val,
-				       props->ptask);
```

Comment:
I'm bit worried on vos tree condensing difference caused by this change. For key punch, aggregation will be able to delete the key and whole evtree, for extent punch, aggregation will have to keep the key and evtree with a punch record. The difference could be non-negligible when truncating a huge file (or many files).

I didn't see how will application use this 'last object write' returned by key query, maybe you'd explain it a bit first?

Reply:
Yeah, that is the downside of this change.   Here is the issue we are trying to solve.

We have learned that a customer has some application for something like a periodic backup and they use the mtime to determine whether a given file has been modified.   I thought about a few different options to track this and settled initially on the timestamp cache.  But the problem is entries can get evicted or the whole cache can be deallocated for things like container close so I couldn't think of a good way to mitigate this and provide what they want.  I don't think we want to return latest epoch for containers that are newly opened as it will make it look like every file has changed.

So, this seemed like a reasonable tradeoff.   The problem is that if the dkey gets punched and removed, the object may also get removed and we lose what may be the only information to indicate the latest timestamp for the file.  So that is why I changed it, specifically to avoid the cleanup.   On the bright side, it should generally be fairly minimal space overhead but I do agree that it's overhead.

Reply:
I see, thanks. It looks to me the best way is to retain object (which will have less overhead), but I don't know if it's easy to implement that.